packages/kotti-input-number/src/InputNumber.vue

```diff
@@ -72,35 +66,72 @@ export default {
 		increaseButtonStyle() {
 			return {
 				'kt-input-number__button': true,
-				'kt-input-number__button--disabled':
-					this.currentValue == this.max || this.disabled,
+				'kt-input-number__button--disabled': this.incrementDisabled,
 			}
 		},
 		decreaseButtonStyle() {
 			return {
 				'kt-input-number__button': true,
-				'kt-input-number__button--disabled':
-					this.currentValue == this.min || this.disabled,
+				'kt-input-number__button--disabled': this.decrementDisabled,
 			}
 		},
 	},
 	watch: {
 		value(val) {
 			this.currentValue = parseFloat(val)
 		},
+		currentValue(newVal) {
+			if (!this.formError) {
+				this.$emit('input', newVal)
+			}
+			if (typeof this.min === 'number') {
+				if (newVal - this.step < this.min) {
+					this.decrementDisabled = true
+				} else if (!this.disabled) {
+					this.decrementDisabled = false
+				}
+			}
+			if (typeof this.max === 'number') {
+				if (newVal + this.step > this.max) {
+					this.incrementDisabled = true
+				} else if (!this.disabled) {
+					this.incrementDisabled = false
+				}
+			}
+		},
+		disabled(newVal) {
+			if (newVal) {
+				this.incrementDisabled = newVal
+				this.decrementDisabled = newVal
+			}
+		},
+	},
+	created() {
+		this.currentValue = typeof this.value === 'number' ? this.value : 0
 	},
 	methods: {
+		incrementValue() {
+			if (this.incrementDisabled) return
+			this.currentValue += this.step
+		},
+		decrementValue() {
+			if (this.decrementDisabled) return
+			this.currentValue -= this.step
+		},
 		/* Do not use input.stepUp - this method does not work in internet explorer and is not transpiled by Babel  */
+
 		setCurrentValue(value) {
 			if (value === this.currentValue) return
-			this.currentValue = typeof value === Number ? value : parseFloat(value)
-			this.$emit('input', this.currentValue)
+			this.currentValue =
```

Comment:
yes, i intentionally did this. I used small modular components because i had to add guard. And the reason I had to use guards is that computed is not functioning as expected.

Reply:
also regarding the watcher - i have to validate that the user did not input a false value before emitting - the component react by showing its invalid, but state should not be updated externally.

Reply:
i personally prefer watchers to computed bc i've run into a case when computed didn't seem like it behaved correctly as well with a reactive property

don't remove the watcher, i edited my comment, just move the logic of currentValue watcher to the setter and make the watcher call the setter

this way you will always set after running all the needed checks 
and always use the setter when u assign a new value to the currentValue